openapi/src/main/scala/endpoints/openapi/OpenApi.scala

```diff
@@ -0,0 +1,131 @@
+    ObjectEncoder.instance { openApi =>
+
+      def jsonMediaTypes(mediaTypes: Map[String, MediaType]): Json =
+        Json.fromFields(mediaTypes.map { case (tpe, mediaType) => tpe -> Json.obj() /* TODO */ })
+
+      JsonObject.fromMap(Map(
+        "openapi" -> Json.fromString("3.0.0"),
+        "info" -> Json.obj(
+          "title" -> Json.fromString(openApi.info.title),
+          "version" -> Json.fromString(openApi.info.version)
+        ),
+        "paths" -> Json.fromFields(
+          openApi.paths.to[List].map { case (path, item) =>
+            val itemObj =
+              Json.fromFields(
+                item.operations.to[List].map { case (verb, op) =>
+                  val fields =
+                    (
+                      "parameters" -> Json.fromValues(
+                        op.parameters.map { parameter =>
+                          val fields =
+                            "name" -> Json.fromString(parameter.name) ::
+                            "in" -> Json.fromString(parameter.in match {
+                              case In.Cookie => "cookie"
+                              case In.Header => "header"
+                              case In.Path   => "path"
+                              case In.Query  => "query"
+                            }) ::
+                            Nil
+                          Json.fromFields(
+                            if (parameter.required) "required" -> Json.fromBoolean(true) :: fields
+                            else fields
+                          )
+                        }
+                      )
+                    ) ::
+                    (
+                      "responses" -> Json.fromFields(
+                        op.responses.to[List].map { case (status, resp) =>
+                          status.toString -> Json.fromFields(
+                            "description" -> Json.fromString(resp.description) ::
+                            (if (resp.content.nonEmpty) {
+                              "content" -> jsonMediaTypes(resp.content) ::
+                              Nil
+                            } else Nil)
+                          )
+                        }
+                      )
+                    ) ::
+                    Nil
+
+                  val fieldsWithRequestEntity =
+                    op.requestBody.map { requestBody =>
+                      "requestBody" -> Json.fromFields({
+                        val requiredFields =
+                          "content" -> jsonMediaTypes(requestBody.content) ::
+                          Nil
+                        requestBody.description.fold(requiredFields)(d => "description" -> Json.fromString(d) :: requiredFields)
+                      })
+                    }.fold(fields)(_ :: fields)
+
+                  verb -> Json.fromFields(fieldsWithRequestEntity)
+                }
+              )
+            (path, itemObj)
+          }
+        )
+      ))
+    }
```

Comment:
Then the case classes hierarchy will become complicated as well :)

I can break down the method into simpler parts.